docs/derivation-rules.md

```diff
@@ -0,0 +1,71 @@
+This file shows derivation rules used in fialyzer.
+
+Our derivation rules are almost same as [the original paper](https://it.uu.se/research/group/hipe/papers/succ_types.pdf)'s one, but extended by remote call, local call, list, etc.
+
+```
+A : context (mapping of variable to type)
+τ : type
+α, β : type variable
+x : variable
+e : term
+fun f/a : local function (f: function name, a: arity)
+fun m:f/a : remote function (m: module name, f: function name, a: arity)
+```
+
+```
+-------------------------- [VAR]
+A ∪ {x → τ} ⊢ x : τ, ∅
+
+A ⊢ e1 : τ1, C1 ... en : τn, Cn
+------------------------------------------------------ [STRUCT]
+A ⊢ {e1, ... , en} : {τ1, ... , τn}, C1 ∧ ... ∧ Cn
+
+A ⊢ e1 : τ1, C1         A ∪ {x → τ1} ⊢ e2 : τ2, C2
+---------------------------------------------------- [LET]
+A ⊢ let x = e1 in e2 : τ2, C1 ∧ C2
+
+A ∪ {fun xi/ai → αi} ⊢ f1 : τ1, C1 ... fn : τn, Cn      e : τ, C   where ai = length (args fi)
+------------------------------------------------------------------------------------------------- [LETREC]
+A ⊢ letrec x1 = f1, ... , xn = fn in e : τ, C1 ∧ ... Cn ∧ C ∧ (τ1 = α1) ∧ ... ∧ (τn = αn)
+
+A ∪ {x1 → α1, ... , xn → αn} ⊢ e : τ, C
+---------------------------------------------------------------------- [ABS]
+A ⊢ fun(x1, ... , xn) → e : (α1, ... , αn) → τ, C
+
+A ⊢ e : τ, C  e1 : τ1, C1 ... en : τn, Cn
+-------------------------------------------------------------------------------------------------------------- [APP]
+A ⊢ e(e1, ... , en) : β, (τ = (α1, ... , αn) → α) ∧ (β ⊆ α) ∧ (τ1 ⊆ α1) ∧ ... ∧ (τn ⊆ αn) ∧ C ∧ C1 ∧ ... ∧ Cn
+
+A ⊢ p : τ, Cp     A ⊢ g : boolean(), Cg
+------------------------------------------ [PAT]
+A ⊢ p when g : τ, Cp ∧ Cg
```

Comment:
In the our implementation of the derivation, a subtyping constraint of the guard expressions is added.
```ocaml
A ⊢ p : τ, Cp     A ⊢ g : τg, Cg
------------------------------------------ [PAT]
A ⊢ p when g : τ, τg ⊆ boolean() ∧ Cp ∧ Cg
```

Reply:
You are right. Thanks!

Reply:
fixed https://github.com/dwango/fialyzer/pull/164/commits/c7aeb50aefc05d4bcaeb4cfd69246df0603c9d94